pkg/controllers/proxy/service_endpoints_sync.go

```diff
@@ -548,6 +548,23 @@ func (nsc *NetworkServicesController) cleanupStaleIPVSConfig(activeServiceEndpoi
 					ipvsServiceString(ipvsSvc), err.Error())
 				continue
 			}
+
+			labelValues, ok := nsc.metricsMap[key]
```

Comment:
I'm always leary about storing string states in the control loop structure as it has the potential to bloat quickly on some cluster edge cases and then cause memory leaks.

In this case, I think that your use here is justified, as you have access to some information on publish that you don't have here. That being said, I'd like to be more careful that this map never has the chance to grow unbound. So I'd like to see two changes.

1. Delete the items in the map when you're done with them (i.e. in this step)
2. I see a small edge case here that the loop that you're embedding this into may not accurately reflect the current state of `nsc.metricsMap`. This loop is based upon the current ipvs services that it found on the node. However, something could happen where the map and the set of current ipvs services on the node become different and there may be some entries in the map that won't ever get cleaned. So I think it would be better if this logic was in its own loop.

Instead of nesting this here, abstract this to its own function, with `activeServiceEndpointMap` as a parameter? Then you could call it from `syncIpvsServices()` directly after `cleanupStaleVIPs()`. In this function you would loop over `nsc.metricsMap`, look for a matching entry in `activeServiceEndpointMap` (the key structures *should* match if I'm reading the code correctly), if you don't find a match, do the `DeleteLabelValues` calls along with deleting the `nsc.metricsMap` entry as well.

Reply:
Good catch on the map key deletion.

I can move this out and call as you suggest, but a lot of this function will need to get duplicated there in order to get the services, check excluded cidrs, derive the key for the service, etc. Are you sure that's what you want?

If it is what you want, then it seems to me that the new function should be called directly after `cleanupStaleIPVSConfig` or else the metrics would be deleted too soon (i.e. potentially before the service itself is actually deleted). WDYT?

Reply:
Maybe you're seeing something that I'm not, but I think you can just do something like:
```
if !nsc.MetricsEnabled {
  return
}

for key, values := range nsc.metricsMap {
  if _, ok := activeServiceEndpointMap[key]; !ok {
    <delete metric labels and entry from map>
  }
}
```

This should work because the key structure of your map and `activeServiceEndpointMap` should be the same: https://github.com/cloudnativelabs/kube-router/blob/master/pkg/controllers/proxy/service_endpoints_sync.go#L120 https://github.com/cloudnativelabs/kube-router/blob/master/pkg/controllers/proxy/service_endpoints_sync.go#L145

And you're right, it should be after `cleanupStaleIPVSConfig` sorry, I got that wrong on my previous comment.

Reply:
I think you're right about the implementation; I was thinking that we'd have to handle iterating over the IPVS services and the differences with FWMark, but I agree with your approach. New commit is coming up.